File: DailyChallenges/CPP/LeetCode_502.cpp
class Solution {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        int n=profits.size();
        
        vector<pair<int,int>> projects(n);
        for(int i=0;i<n;i++) projects[i]={profits[i],capital[i]}; 
        sort(projects.begin(),projects.end(),[&](auto a, auto b){ return a.second<b.second;});
        
        priority_queue<int> maxProfit;
        int i=0;
        while(k--){
            while(i<n && w>=projects[i].second) 
                maxProfit.push(projects[i++].first);
            
            if(!maxProfit.empty()){ 
                w+=maxProfit.top();
                maxProfit.pop();
            }
        }
        return w;
    }
};

File: DailyChallenges/Scala/LeetCode_502.scala
object Solution {
    def findMaximizedCapital(k: Int, w: Int, profits: Array[Int], capital: Array[Int]): Int = {
        val projects = (profits zip capital).sortBy(_._2)
        val pq = collection.mutable.PriorityQueue.empty[Int]
        var idx = 0
        def loop(mw: Int, cnt: Int): Int = if (cnt > k) mw else {
            while (idx < projects.length && projects(idx)._2 <= mw) {
                pq.enqueue(projects(idx)._1)
                idx += 1
            }
            if (pq.isEmpty) mw else loop(mw + pq.dequeue, cnt+1)
        }
        loop(w, 1)
    }
}
